File: src/leetcode/python/hard/CopyListWithRandomPointer_138.py
class RandomListNode(object):
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None


class CopyListWithRandomPointer_138(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:
            return None
        node = head
        table = {}
        while node:
            table[node] = RandomListNode(node.label)
            node = node.next
        node = head
        while node:
            if node.next:
                table[node].next = table[node.next]
            if node.random:
                table[node].random = table[node.random]
            node = node.next
        return table[head]

File: src/leetcode/java/hard/CopyListWithRandomPointer_138.java
package leetcode.java.hard;

import java.util.HashMap;

public class CopyListWithRandomPointer_138 {
  public RandomListNode copyRandomList(RandomListNode head) {
    HashMap<RandomListNode, RandomListNode> table = new HashMap<>();
    RandomListNode node = head;
    while (node != null) {
      table.put(node, new RandomListNode(node.label));
      node = node.next;
    }
    node = head;
    while (node != null) {
      table.get(node).next = table.get(node.next);
      table.get(node).random = table.get(node.random);
      node = node.next;
    }
    return table.get(head);
  }

  class RandomListNode {
    int label;
    RandomListNode next, random;

    RandomListNode(int x) {
      this.label = x;
    }
  }
}
